New MCP Chat Client Plugin

Lucifergene opened the pull request:
Signed-off-by: Avik Kundu [akundu@redhat.com](mailto:akundu@redhat.com)

### Introducing the MCP Chat Client Plugin!

The MCP Chat Client plugin brings conversational AI capabilities directly into your Backstage environment. It leverages the Model Context Protocol to connect with various AI providers and external tools, enabling developers to interact with their infrastructure, catalogs, and external services through natural language.

**Features:**
-  **Multi-Provider AI Support:** Works with OpenAI, Claude, Gemini, and Ollama
-  **Multi-Server Support:** Connect multiple MCP servers (STDIO, SSE, Streamable HTTP)
- **Tool Management:** Browse and dynamically enable/disable tools from connected MCP servers
- **Rich Chat Interface:** Beautiful, responsive chat UI with markdown support
- **Quick Setup:** Configurable QuickStart prompts for common use cases

**Screenshots:**
<div align="center">
<table>
  <tr>
    <td style="padding: 0 10px;">
      <img src="https://github.com/Lucifergene/backstage-community-plugins/raw/mcp-chat-client/workspaces/mcp-chat/docs/images/quick-prompts.png" alt="Quick Prompts" width="600"/>
      <div align="center"><em>Pre-configured prompts for common tasks</em></div>
    </td>
    <td rowspan="2" style="vertical-align: top; padding: 0 10px;">
      <img src="https://github.com/Lucifergene/backstage-community-plugins/raw/mcp-chat-client/workspaces/mcp-chat/docs/images/mcp-tools-panel.png" alt="MCP Tools Panel" height="500"/>
      <div align="center"><em>Available MCP tools and server connections</em></div>
    </td>
  </tr>
  <tr>
    <td style="padding: 0 10px;">
      <img src="https://github.com/Lucifergene/backstage-community-plugins/raw/mcp-chat-client/workspaces/mcp-chat/docs/images/chat-interface.png" alt="Chat Interface" width="600"/>
      <div align="center"><em>The main chat interface with AI responses and tool integration</em></div>
    </td>
  </tr>
</table>
</div>

**Video Walkthough:** 
https://drive.google.com/file/d/17_0vZFd6txRJ1_Mioc7rUXYOwBs546Zs/view?usp=sharing

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/community-plugins/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/community-plugins/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

vinzscam commented on workspaces/mcp-chat-client/plugins/mcp-chat-client-backend/src/router.ts:
```diff
@@ -0,0 +1,304 @@
+      const { messages, enabledTools = [] } = req.body;
+
+      // Validate messages array
+      if (!messages || !Array.isArray(messages) || messages.length === 0) {
+        return res.status(400).json({ error: 'No query provided' });
+      }
```
this needs some deep validation

vinzscam commented on workspaces/mcp-chat-client/plugins/mcp-chat-client-backend/src/router.ts:
```diff
@@ -0,0 +1,304 @@
+        return res.status(400).json({ error: 'No query provided' });
+      }
+
+      const serverConfigs = loadServerConfigs(config);
+      await mcpClientService.initMCP(serverConfigs);
```
shouldn't these two steps be invoked at startup instead?

Lucifergene replied:
makes sense 👍

vinzscam commented on workspaces/mcp-chat-client/plugins/mcp-chat-client-backend/src/router.ts:
```diff
@@ -0,0 +1,304 @@
+  });
+
+  // Test route with fixed prompt
+  router.get('/test/latest-news', async (_req, res) => {
```
can all `/test/*` routes be removed?

Lucifergene replied:
Done, have removed this endpoint and renamed other according to their functions

vinzscam commented on workspaces/mcp-chat-client/plugins/mcp-chat-client-backend/src/services/MCPClientServiceImpl.ts:
```diff
@@ -0,0 +1,349 @@
+export class MCPClientServiceImpl implements MCPClientService {
+  private readonly logger: LoggerService;
+  private readonly config: RootConfigService;
+  private llmProvider!: LLMProvider;
```
perhaps `initializeLLMProvider()` could return the llmProvider so that we remove the non null assertion here

vinzscam commented on workspaces/mcp-chat-client/plugins/mcp-chat-client-backend/src/services/MCPClientServiceImpl.ts:
```diff
@@ -0,0 +1,349 @@
+      }
+
+      // Connect the client with the appropriate transport
+      await client.connect(transport);
```
if one of the servers is down would it prevent the plugin to work with other servers?

vinzscam commented on workspaces/mcp-chat-client/plugins/mcp-chat-client-backend/src/services/MCPClientServiceImpl.ts:
```diff
@@ -0,0 +1,349 @@
+        toolCalls: [],
+        toolResponses: [],
+      };
+    } catch (error) {
+      throw error;
+    }
```
this try catch can be removed

Lucifergene replied:
Done

vinzscam commented on workspaces/mcp-chat-client/plugins/mcp-chat-client-backend/src/services/MCPClientServiceImpl.ts:
```diff
@@ -0,0 +1,349 @@
+    toolCalls: ToolCall[];
+    toolResponses: any[];
+  }> {
+    const messages: ChatMessage[] = [
```
won't this logic cause the initial prompt to be replicated multiple times while the conversation is going?

vinzscam commented on workspaces/mcp-chat-client/plugins/mcp-chat-client-backend/src/services/MCPClientServiceImpl.ts:
```diff
@@ -0,0 +1,349 @@
+      {
+        role: 'system',
+        content:
+          "You are a helpful assistant. When using tools, provide a clear, readable summary of the results rather than showing raw data. Focus on answering the user's question with the information gathered.",
```
is this a golden prompt that makes everything work or should be customizable by the adopters?

Lucifergene replied:
Yes, this is a golden prompt that works across all use cases.
Based on all my tests so far, it has consistently delivered results. In fact, it worked so well that I had completely forgotten about it until now!

vinzscam commented on workspaces/mcp-chat-client/plugins/mcp-chat-client/src/api/McpChatApi.ts:
```diff
@@ -0,0 +1,182 @@
+    });
+
+    if (!response.ok) {
+      throw new Error(`Chat request failed: ${response.statusText}`);
```
same below 
```suggestion
      throw await ResponseError.fromResponse(response);
```

Lucifergene replied:
Thanks for this! Something new to know!
Have updated all the required ones

vinzscam commented on workspaces/mcp-chat-client/plugins/mcp-chat-client/src/components/RightPane/ActiveMcpServers.tsx:
```diff
@@ -0,0 +1,152 @@
+    return 'transparent';
+  };
+
+  const getChipColor = (server: MCPServer) => {
```
use material-ui theming for all these colors so that the plugin works with custom themes. Same in other files

vinzscam commented on workspaces/mcp-chat-client/plugins/mcp-chat-client/src/components/RightPane/RightPane.tsx:
```diff
@@ -0,0 +1,329 @@
+  });
+
+  // Fetch available tools when component mounts or servers change
+  useEffect(() => {
```
use `useAsync` when is possible for all the async state changes: The code gets more compact and we are safer if the component unmounts while async operations are in progress

Lucifergene replied:
Thanks, it was something new for me. Explored the "react-use" library, its a great one!

vinzscam commented on workspaces/mcp-chat-client/plugins/mcp-chat-client/src/api/McpChatApi.ts:
```diff
@@ -0,0 +1,182 @@
+export interface ConfigStatus {
+  provider: any;
+  mcpServers: Array<{
+    id?: string;
```
Is this correct? It looks non optional in the config definition

vinzscam commented on workspaces/mcp-chat-client/README.md:
```diff
@@ -0,0 +1,378 @@
+### Development Guidelines
+
+- Follow the existing code style and patterns
+- Add tests for new functionality
```
we recommend adding tests but there are no tests at the moment 😅

vinzscam commented on workspaces/mcp-chat-client/plugins/mcp-chat-client-backend/src/providers/base-provider.ts:
```diff
@@ -0,0 +1,114 @@
+
+export interface Tool {
+  type: 'function';
+  function: {
```
this `function` property is a bit ambiguous 😅

Lucifergene commented on workspaces/mcp-chat-client/plugins/mcp-chat-client-backend/src/router.ts:
```diff
@@ -0,0 +1,304 @@
+      }
+    }
+
+    let env: Record<string, string> | undefined;
```
Sorry, missed that. Have added the missing env as well as args there

Lucifergene commented on workspaces/mcp-chat-client/plugins/mcp-chat-client-backend/src/router.ts:
```diff
@@ -0,0 +1,304 @@
+        mcpServers: serverList,
+      });
+    } catch (error) {
+      res.status(500).json({
```
Removed 👍

Lucifergene commented:
Thanks a lot, @vinzscam, for the detailed initial review 🚀

I’ve addressed all your comments and learned some new optimisation techniques for Backstage plugins along the way.

Additionally, I’ve added extensive unit tests for the key files.

The PR is now ready for the second round of reviews 👍

awanlin commented:
HI @Lucifergene, just following up on your item on the Community Plugins SIG agenda, we generally don't like to review PRs during the SIG. That really should happen here. A few things worth mentioning though: it's summer time so people are away on vacation and things will be slower, also this is a volunteer based Community which means people are doing these reviews above and beyond their normal work. Appreciate your patience as you go through the review process 🙏

Lucifergene commented:
@awanlin Any updates on this PR? It has been a month since our last communication.

modethirteen commented:
I'm very intrigued by this effort. Is this a backport of a capability that Red Hat developed for Red Hat Developer Hub?

christoph-jerolimov commented on workspaces/mcp-chat/plugins/mcp-chat-backend/src/providers/claude-provider.ts:
```diff
@@ -0,0 +1,193 @@
+          if (errorData.error?.message) {
+            errorMessage = errorData.error.message;
+          }
+        } catch {
+          // If parsing fails, use the raw error text but truncate if too long
+          errorMessage =
+            errorText.length > 100
+              ? `${errorText.substring(0, 100)}...`
+              : errorText;
+        }
+
+        // Provide user-friendly messages for common errors
+        if (response.status === 401) {
+          errorMessage =
+            'Invalid API key. Please check your Claude API key configuration.';
+        } else if (response.status === 429) {
+          errorMessage =
+            'Rate limit exceeded. Please try again later or check your Claude usage limits.';
+        } else if (response.status === 403) {
+          errorMessage =
+            'Access forbidden. Please check your API key permissions.';
+        }
+
+        return {
+          connected: false,
+          error: errorMessage,
+        };
```
Not critical, but maybe an idea for the future. If you return the status code and origin message to the frontend, couy can truncate and translate these strings there easier.

Not necessary to me to change this right now.

christoph-jerolimov commented on workspaces/mcp-chat/plugins/mcp-chat-backend/src/providers/openai-provider.ts:
```diff
@@ -0,0 +1,119 @@
+          if (errorData.error?.message) {
+            errorMessage = errorData.error.message;
+          }
+        } catch {
+          // If parsing fails, use the raw error text but truncate if too long
+          errorMessage =
+            errorText.length > 100
+              ? `${errorText.substring(0, 100)}...`
+              : errorText;
+        }
+
+        // Provide user-friendly messages for common errors
+        if (response.status === 401) {
+          errorMessage =
+            'Invalid API key. Please check your OpenAI API key configuration.';
+        } else if (response.status === 429) {
+          errorMessage =
+            'Rate limit exceeded. Please try again later or check your OpenAI usage limits.';
+        } else if (response.status === 403) {
+          errorMessage =
+            'Access forbidden. Please check your API key permissions.';
+        }
+
```
Same here. I like to recommend to do truncation and error messages in the frontend.

But no necessary change for now.

christoph-jerolimov commented on workspaces/mcp-chat/plugins/mcp-chat-backend/src/services/MCPClientServiceImpl.ts:
```diff
@@ -0,0 +1,471 @@
+
+        let transport;
+
+        if (serverConfig.type === 'streamable-http') {
```
I think its worth to create an enum for the type.

christoph-jerolimov commented on workspaces/mcp-chat/plugins/mcp-chat-backend/src/services/MCPClientServiceImpl.ts:
```diff
@@ -0,0 +1,471 @@
+                }. Please ensure Node.js is properly installed with npx available.`,
+              );
+            }
+          } else if (serverConfig.scriptPath) {
+            // Use script path
+            const isPythonScript = serverConfig.scriptPath.endsWith('.py');
+            const isWindows = process.platform === 'win32';
+
+            if (isPythonScript) {
+              command = isWindows ? 'python' : 'python3';
+            } else {
+              command = process.execPath;
+            }
+            args = [serverConfig.scriptPath, ...(serverConfig.args || [])];
```
I think there is no documentation around this?

I also not sure if this magic is needed or if you can just people configure:

```yaml
  - id: local-script-server
    name: Local Script Server
    command:
      - python3
      - ./servers/my-mcp-server.js
      - --port=9100
      - --env=dev
```

So this isn't wrong. I just couldn't find that in the READMEs or SERVER_CONFIGURATION.md

christoph-jerolimov commented on workspaces/mcp-chat/plugins/mcp-chat-backend/src/services/MCPClientServiceImpl.ts:
```diff
@@ -0,0 +1,471 @@
+            command,
+            args,
+            env: {
+              ...process.env, // Inherit current environment
```
This makes it really handy for developers, but I don't know if it would be "saver" if you don't pass-though all environment variables. But I'm not an expert here.

christoph-jerolimov commented on workspaces/mcp-chat/plugins/mcp-chat-backend/src/utils.ts:
```diff
@@ -0,0 +1,405 @@
+  // Get the directory where node is installed
+  const nodeDir = path.dirname(process.execPath);
+
+  const possiblePaths = [
+    'npx', // Try system PATH first
+    path.join(nodeDir, 'npx'), // Same dir as node (Unix)
+    path.join(nodeDir, 'npx.cmd'), // Windows
+    '/usr/local/bin/npx', // Common installation path
+    '/opt/homebrew/bin/npx', // Homebrew on Apple Silicon
+  ];
```
My opinion: People should just configure their system correct. I would avoid this kind of "magic", but its up to you. :)

christoph-jerolimov commented on workspaces/mcp-chat/plugins/mcp-chat/src/components/BotIcon/BotIcon.test.tsx:
```diff
@@ -0,0 +1,222 @@
+
+    it('handles multiple props together', () => {
+      const { container } = renderWithTheme(
+        <BotIcon size={64} color="#blue" className="test-icon" />,
+      );
+
+      const icon = container.querySelector('svg');
+      expect(icon).toHaveAttribute('width', '64');
+      expect(icon).toHaveAttribute('height', '64');
+      expect(icon).toHaveAttribute('fill', '#blue');
```
doesn't matter, but the color should be a string or hex code right?

```suggestion
        <BotIcon size={64} color="blue" className="test-icon" />,
      );

      const icon = container.querySelector('svg');
      expect(icon).toHaveAttribute('width', '64');
      expect(icon).toHaveAttribute('height', '64');
      expect(icon).toHaveAttribute('fill', 'blue');
```

christoph-jerolimov commented on workspaces/mcp-chat/plugins/mcp-chat/src/components/ChatPage/ChatPage.tsx:
```diff
@@ -0,0 +1,129 @@
+                />
+              </Box>
+            ) : (
+              <>
+                {/* Chat Container */}
+                <ChatContainer
+                  ref={chatContainerRef}
+                  sidebarCollapsed={sidebarCollapsed}
+                  mcpServers={mcpServers}
+                  messages={messages}
+                  setMessages={setMessages}
+                />
+
+                {/* Sidebar - Right Side */}
+                <RightPane
```
My you could use a drawer instead, but your solution is working fine :+1:

https://v5.mui.com/material-ui/react-drawer/#mini-variant-drawer

christoph-jerolimov commented on workspaces/mcp-chat/plugins/mcp-chat/src/components/RightPane/RightPane.tsx:
```diff
@@ -0,0 +1,253 @@
+  return (
+    <Box
+      sx={{
+        width: sidebarCollapsed ? 60 : 400,
```
This is the reason why the sidebar animates also the content. I think it might be an option to replace the custom sidebar with a MUI Drawer. But your solution is working so no need to touch it now!

https://v5.mui.com/material-ui/react-drawer/#mini-variant-drawer

christoph-jerolimov reviewed (commented):
Thanks a lot @Lucifergene for this really nice looking MCP Chat experience in Backstage.

I've tested the plugin with the included dev page and it works nice. I have below a bunch of small suggestions but none of them should keep your first release back.

Here are some general recommendations, but at the end this is up to you:

1. Maybe you should release a 0.1 first and test the released npm packages once. I don't want hold you back from a 1.0 release, but for me, its always how I would try that because it happens quickly to have a typo in one of the exports or readmes.
2. As I shared, I couldn't verify the plugin yet in a Backstage app, but maybe that's easier with a released version.
3. I like to recommend you to add also a security node about mcp servers and that people should just include trusted mcp servers.

My tests with the demo app:

https://github.com/user-attachments/assets/03a4d4f7-052c-4eec-97fa-07de34e5972e

Lucifergene commented:
Fixing the review comments soon

Lucifergene commented on workspaces/mcp-chat/plugins/mcp-chat/src/components/BotIcon/BotIcon.test.tsx:
```diff
@@ -0,0 +1,222 @@
+  return render(<ThemeProvider theme={theme}>{component}</ThemeProvider>);
+};
+
+describe('BotIcon', () => {
```
No I guess, have removed that test file.

Lucifergene commented on workspaces/mcp-chat/plugins/mcp-chat/src/components/ChatContainer/ChatContainer.test.tsx:
```diff
@@ -0,0 +1,537 @@
+      expect(screen.getByTestId('SendIcon')).toBeInTheDocument();
+    });
+
+    it('adjusts layout based on sidebar state', () => {
+      const { rerender } = renderChatContainer({ sidebarCollapsed: true });
+
+      rerender(
+        <TestApiProvider apis={[[mcpChatApiRef, mockMcpChatApi]]}>
+          <ThemeProvider theme={createTheme()}>
+            <ChatContainer {...defaultProps} sidebarCollapsed={false} />
+          </ThemeProvider>
+        </TestApiProvider>,
+      );
+
+      expect(
+        screen.getByPlaceholderText('Message Assistant...'),
+      ).toBeInTheDocument();
+    });
```
Have reduced the tests in the file

Lucifergene commented on workspaces/mcp-chat/plugins/mcp-chat/src/components/ChatContainer/ChatContainer.tsx:
```diff
@@ -0,0 +1,338 @@
+  sidebarCollapsed: boolean;
+  mcpServers: MCPServer[];
+  messages: Message[];
+  setMessages: React.Dispatch<React.SetStateAction<Message[]>>;
```
Thanks, it is better

Lucifergene commented on workspaces/mcp-chat/plugins/mcp-chat/src/components/ChatContainer/QuickStart.test.tsx:
```diff
@@ -0,0 +1,284 @@
+  });
+
+  describe('Interaction Handling', () => {
+    it('accepts onSuggestionClick prop', () => {
```
updated the tests with meaningful ones now

Lucifergene commented on workspaces/mcp-chat/plugins/mcp-chat/src/components/ChatContainer/QuickStart.test.tsx:
```diff
@@ -0,0 +1,284 @@
+      expect(screen.getByText('Code Analysis')).toBeInTheDocument();
+    });
+
+    it('handles missing onSuggestionClick gracefully', () => {
+      expect(() => {
+        renderWithProviders(
+          <QuickStart onSuggestionClick={undefined as any} />,
+          mockConfigApiWithSuggestions,
+        );
+      }).not.toThrow();
+    });
```
updated the tests with meaningful ones now

drodil commented on workspaces/mcp-chat/plugins/mcp-chat-backend/src/services/MCPClientServiceImpl.ts:
```diff
@@ -0,0 +1,471 @@
+      messages.unshift({
+        role: 'system',
+        content:
+          "You are a helpful assistant. When using tools, provide a clear, readable summary of the results rather than showing raw data. Focus on answering the user's question with the information gathered.",
```
Just an idea, would it make sense to be able to configure the system prompt using `Config`?

Lucifergene replied:
Thanks, have implemented it.
You can now set this directly from the config.

Lucifergene commented on workspaces/mcp-chat/plugins/mcp-chat/src/components/RightPane/ActiveMcpServers.tsx:
```diff
@@ -0,0 +1,184 @@
+              clickable={server.status?.connected}
+              onClick={
+                server.status?.connected
+                  ? () => onServerToggle(server.name)
```
Yeah, the IDs are unique, have modified that code

Lucifergene commented on workspaces/mcp-chat/plugins/mcp-chat/src/components/RightPane/ActiveMcpServers.tsx:
```diff
@@ -0,0 +1,184 @@
+      <Box sx={{ display: 'flex', flexWrap: 'wrap' }}>
+        {mcpServers.map(server => (
+          <Tooltip
+            key={server.name}
```
Yeah, the IDs are unique, have modified that code

drodil commented on workspaces/mcp-chat/plugins/mcp-chat-backend/src/services/MCPClientServiceImpl.ts:
```diff
@@ -0,0 +1,471 @@
+      }
+    }
+
+    this.tools = allTools;
```
I think there might be issue that the tools are only updated on the plugin startup and not after that.. What if some MCP server adds or removes some tool during runtime?  It's the same issue with scaffolder at the moment, see this: https://github.com/backstage/backstage/pull/31244

Lucifergene replied:
You're right that tools are only fetched when the plugin starts up. However, this is actually the expected behavior for MCP servers. In practice, MCP servers define their available tools as part of their initial connection handshake and don't typically add or remove tools while running. The tools are part of the server's static capabilities, so they shouldn't change during runtime.

If someone does need to refresh the tool list (maybe after updating an MCP server or restarting it), they can just restart the Backstage backend or toggle the server off and on in the UI. 

We could add a manual refresh button in the future if this becomes a common need, but for now, the current approach keeps things simple and avoids unnecessary polling overhead.

drodil commented on workspaces/mcp-chat/plugins/mcp-chat-backend/src/services/MCPClientServiceImpl.ts:
```diff
@@ -0,0 +1,471 @@
+      return this.mcpServers ? await this.mcpServers : [];
+    }
+
+    const allTools: ServerTool[] = [];
```
You could add tools from the backstage `actionsRegistry` directly here?

drodil reviewed (commented):
This is very interesting, but a question: how does the user authentication work towards the MCP servers? Backstage MCP backend provides dynamic client registration, but based on a quick look at the code, it doesn't have support for that?

Which made me wonder, would it make sense to provide the backstage actions registry tools directly to the client to use? That way, it would use the user's credentials to run the tools inside the backstage instance.

Lucifergene commented on workspaces/mcp-chat/plugins/mcp-chat/package.json:
```diff
@@ -0,0 +1,83 @@
+{
+  "name": "@backstage-community/plugin-mcp-chat",
+  "description": "A Backstage plugin that provides a chat interface for interacting with the MCP Servers.",
+  "version": "0.1.0",
```
Changed to 0.0.0

Lucifergene commented on workspaces/mcp-chat/plugins/mcp-chat/README.md:
```diff
@@ -0,0 +1,361 @@
+
+### Frontend Installation
+
+1. **Install the frontend plugin**:
```
Sure, we are planning to add the support in the next release.

Lucifergene commented:
@vinzscam Please take a look. I believe it’s ready to merge now. We can add the new features, like the New Frontend system and other requested enhancements, in subsequent releases.

Lucifergene commented:
Thanks @dfl-aeb, I was trying to find the package in NPM, but couldn't.
I am trying to figure out the issue.
Any help would be appreciated.

dfl-aeb commented:
Copilot to the rescue 🆘

I assume it's related to this: 
@internal/mcp-chat lacks `"private": true` in its [package.json](https://github.com/backstage/community-plugins/blob/main/workspaces/mcp-chat/package.json#L4), so the release workflow attempted to publish it.  
Existing workspaces (octopus-deploy, azure-devops, 3scale, …) all have `"private": true`.

https://github.com/backstage/community-plugins/blob/main/workspaces/octopus-deploy/package.json#L4

The publish then fails with YN0035 (404) because the @internal scope isn’t publishable with the current npm credentials.

So I guess adding `"private": true` to your package.json should fix it?
